File: part3/java/async/AsyncClient.java
import java.io.IOException;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.transport.TNonblockingSocket;
import org.apache.thrift.TException;
import org.apache.thrift.async.TAsyncMethodCall;
import org.apache.thrift.async.TAsyncClientManager;
import org.apache.thrift.async.AsyncMethodCallback;

import TradeReporting.TradeHistory.AsyncClient.get_last_sale_call;

public class AsyncClient {

    private static abstract class WaitableCallback<T extends TAsyncMethodCall> 
            implements AsyncMethodCallback<T> {
        private CountDownLatch latch = new CountDownLatch(1);

        public void reset() {
            latch = new CountDownLatch(1);
        }

        public void complete() {
            latch.countDown();
        }

        public boolean await(int i) {
            boolean b = false;
            try {
                b = latch.await(1000, TimeUnit.MILLISECONDS);
            } catch(Exception e) {
                System.out.println("Await error");
            }
            return b;
        }

        @Override
        public void onError(Exception ex) {
            if (ex instanceof TimeoutException) {
                System.out.println("[Client] Async call timed out");
            } else {
                System.out.println("[Client] Async call error");
            }
            complete();
        }
    }


    public static void main(String[] args) 
            throws IOException, InterruptedException, TException {
        TAsyncClientManager client_man = new TAsyncClientManager();
        TNonblockingSocket trans_ep = new TNonblockingSocket("localhost", 9090);
        TradeReporting.TradeHistory.AsyncClient client = 
                new TradeReporting.TradeHistory.AsyncClient(new TBinaryProtocol.Factory(), 
                                                            client_man, trans_ep);

        WaitableCallback<get_last_sale_call> wc = new WaitableCallback<get_last_sale_call>() {
            @Override
            public void onComplete(get_last_sale_call cob) {
                try {
                    TradeReporting.TradeReport tr = cob.getResult();
                    System.out.println("[Client] received [" + tr.seq_num + "] " + 
                                       tr.symbol + " : " + tr.size + " @ " + tr.price);
                } catch (TException e) {
                    ;
                } finally {
                    complete();
                }
            }
        };

        wc.reset();
        client.get_last_sale("IBM", wc);
        wc.await(1000);

        client.setTimeout(1000);

        wc.reset();
        client.get_last_sale("GE", wc);
        wc.await(1000);

        client_man.stop();
        trans_ep.close();
    }
}

File: part3/cpp/async/async_client.cpp
#include "gen-cpp/helloSvc.h"
#include <thrift/transport/TSocket.h>
#include <thrift/transport/TBufferTransports.h>
#include <thrift/protocol/TBinaryProtocol.h>
#include <memory>
#include <iostream>
#include <future>
#include <chrono>

using namespace apache::thrift::transport;
using namespace apache::thrift::protocol;
using std::shared_ptr;
using std::make_shared;

int main() {
    shared_ptr<TTransport> trans;
    trans = make_shared<TSocket>("localhost", 9090);
    trans = make_shared<TBufferedTransport>(trans);
    auto proto = make_shared<TBinaryProtocol>(trans);
    helloSvcClient client(proto);

    try {
        trans->open();
        
        auto fut = std::async(std::launch::async, [&client]{
            std::string out; 
            client.getMessage(out, "world!"); 
            return out;
        });

        std::cout << "Client async call to getMessage() in progress" << std::endl;
        if (fut.wait_for(std::chrono::milliseconds(500)) == std::future_status::ready) {
            std::cout << fut.get() << std::endl;
        } else {
            throw "async call timed out";
        }
    } catch(...) {
        std::cout << "Client caught an exception" << std::endl;
    } 
    trans->close();
}
